src/index.d.ts

```diff
@@ -0,0 +1,158 @@
+declare module 'react-form' {
+  import {
+    ChangeEventHandler,
+    ComponentType,
+    Dispatch,
+    FocusEventHandler,
+    FormEventHandler,
+    HTMLProps,
+    Provider,
+    SetStateAction,
+  } from 'react'
+
+  type Debounce = <T>(fn: (...args: any[]) => T, wait: number) => Promise<T>
+  type ValidatorReturn = string | false | undefined
+  type OptionalPromise<T> = Promise<T> | T
+
+  interface FormMeta {
```

Comment:
Oh yeah, I was trying to think about a good way to allow this to be typed. I've never used `Record`, cool! In your first snippet, trying to extend `C` emits an error, I'm guessing because it's being declared after being extended. What do you think of keeping `FormMeta` an interface with no generics, and instead using `FormMeta & C` whenever it's referenced?

```ts
interface FormMeta {
  error: string | any
  isSubmitting: boolean
  isDirty: boolean
  isSubmitted: boolean
  submissionAttempts: number
  isValid: boolean
  fieldsAreValidating: boolean
  fieldsAreValid: boolean
  canSubmit: boolean
}

type SetFormMeta<C = Record<string, any>> = Partial<FormMeta & C> | ((previousMeta: FormMeta & C) => FormMeta & C)

interface FormInstance<T, C = Record<string, any>> {
  Form: ComponentType<Omit<HTMLProps<HTMLFormElement>, 'onSubmit'>>
  values: T
  meta: FormMeta & C
  ...
}
```

Reply:
Yeah, I absolutely would prefer separating out the "react-form"-meta and "custom meta".

Reply:
I committed a start to this, but then stopped before committing the rest. I ultimately came up with this: https://gist.github.com/BrianMitchL/8e121cfb1633bb16924c81b6c264d88a

My example app that uses these types started breaking, and I'm having a hard time figuring out where. Using four generics everywhere, especially when we're splitting props out and creating field components, makes things really complex. Is there a better way than how I did it in the gist? I did move all of the default values for the generics to the functions. All of the internal interfaces require the generics.

Reply:
I pushed all of the changes in the gist, so looking here at the PR is a better place to look at it 🙂